review-tools/ghlink

```diff
@@ -0,0 +1,280 @@
+              -verbose => 1);
+}
+if ($man) {
+    pod2usage(-exitval => 0,
+              -verbose => 2);
+}
+
+if ($markdown) {
+    $replace = \&markdown
+}
+
+
+my $remotes = `git remote -v`;
+if ($? != 0 or $remotes !~ "openssl\.git") {
+    die "Current directory does not belong to an OpenSSL git repository";
+}
+
+my $curr_branch = `git rev-parse --abbrev-ref HEAD`;
+chomp $curr_branch;
+
+my $prefix = `git rev-parse --show-prefix`;
+chomp $prefix;
+
+
+# some results are cached for efficiency reasons
+
+my %commits;
+my %urls;
+
+sub check_url {
+    my ($match, $revision, $path, $lineno) = @_;
+
+    if (!defined($revision)) {
+        $revision = $curr_branch;
+    }
+
+    my $abbrev_commit = ($permanent) ? "" : "--abbrev-commit";
+
+    if (!defined($commits{$revision})) {
+        my $c = `git rev-list $abbrev_commit -1 $revision -- 2>/dev/null`;
+        chomp $c;
+
+        if ($? != 0) {
+            $commits{$revision} = "";
+        } elsif ($permanent) {
+            # always use the commit id if --permanent was specified
+            $commits{$revision} = $c;
+        } else {
+            # if a branch name was specified, use if it exists remotely
+            # otherwise, use the commit id
+            `git ls-remote --exit-code $repo --heads refs/heads/$revision`;
+            $commits{$revision} = ($? == 0) ? $revision : $c;
+        }
+    }
+
+    $revision = $commits{$revision};
+
+    if (!$revision) {
+        return "";
+    }
+
+    my $gitpath = "$revision:$prefix$path";
+
+    if (!defined($urls{$gitpath})) {
+        # create urls only for objects in the local repository
+        `git rev-parse $gitpath 2>/dev/null`;
+        if ($? != 0) {
+            $urls{$gitpath} = "";
+        } else {
+            $urls{$gitpath} = "$blob/$revision/$prefix$path";
+        }
+    }
+
+    my $url = $urls{$gitpath};
+
+    if (!$url) {
+        return "";
+    }
+
+    return defined($lineno) ? "$url#L$lineno" : "$url";
+}
+
+sub link {
+    my ($match, $revision, $path, $lineno) = @_;
+
+    my $url = check_url($match, $revision, $path, $lineno);
+
+    if (!$url) {
+        return $match;
+    }
+
+    return $url;
+}
+
+sub markdown {
+    my ($match, $revision, $path, $lineno) = @_;
+
+    my $url = check_url($match, $revision, $path, $lineno);
+
+    if (!$url) {
+        return $match;
+    }
+
+    return "[$match]($url)";
+}
+
+if ($list) {
+    # list mode: print github links for all matched locations (and discard the rest)
+    while (<>) {
+        while (/(?:(${regex_revision}):)?(${regex_path}):(${regex_lineno})?:?/g) {
+            my $found = $replace->($&,$1,$2,$3);
+            if ($found ne $&) {
+                print ("$found\n");
+            }
+        }
+    }
+} else {
+    # replace mode: replace all matched locations in the text with github links
+    while (<>) {
+        s/(?:(${regex_revision}):)?(${regex_path}):(${regex_lineno})?:?/$replace->($&,$1,$2,$3)." "/eg;
+        print;
+    }
+}
```

Comment:
I wonder if it would be better not to have to repeat the regexp...

``` perl
while (<>) {
    my $found;
    if (/(?:(${regex_revision}):)?(${regex_path}):(${regex_lineno})?:?/g) {
        $found = $replace->($&,$1,$2,$3);
    }

    if ($list) {
        # list mode: print github links for all matched locations (and discard the rest)
        if ($found ne $&) {
            $_ = "$found\n";
        } else {
            $_ = "";
        }
    } else {
        # replace mode: replace all matched locations in the text with github links
        if (defined($found)) {
            $_ = $`.$found." ".$';
        }
    }
    print:
}
```

This avoids repeating the same regexp twice.

Reply:
You are right, the regular expression should not be defined twice. However, the solution proposed by you sacrifices the power and elegance of the replacement operator for this goal. Also, it does not work with multiple occurrences of the regular expression in the same line anymore. (Even if I replace the `if` with a 'while' loop, because you modify $_.)

But your comment was the motivation for me to search again whether and how the same (compiled) regular expression can be used not only for searching but also for (search&)replacing. This is what I came up with:

```perl
my $regex = qr/(?:(${regex_revision}):)?(${regex_path}):(${regex_lineno})?:?/;

if ($list) {
    # list mode: print github links for all matched locations (and discard the rest)
    while (<>) {
        while (/$regex/g) {
            my $found = $replace->($&,$1,$2,$3);
            if ($found ne $&) {
                print ("$found\n");
            }
        }
    }
} else {
    # replace mode: replace all matched locations in the text with github links
    while (<>) {
        s/$regex/$replace->($&,$1,$2,$3)." "/eg;
        print;
    }
}
```